src/main/java/com/m2m/shopping/Service/Impl/ProductServiceImpl.java

```diff
@@ -31,4 +30,72 @@ public Product save(Product enity) {
     public boolean delete(Product enity) {
         return super.delete(enity);
     }
+
+    @Override
+    public List<Product> findToPage(Boolean existActive, int pageNumber, int pageSize) {
+        return super.findToPage(Product.class,existActive,pageNumber,pageSize);
+    }
+
+    @Override
+    public List<Product> findPageParam( Boolean existActive, int pageNumber, int pageSize, String title, Integer idCategories) {
+        String sql = "select o from Product o where o.title like ?0 and o.categories.id = ?1";
+        return super.findManyPage(Product.class,sql,pageNumber,pageSize,"%"+title+"%", idCategories );
+    }
+
+    @Override
+    public List<Product> findPageParam(Boolean existActive, int pageNumber, int pageSize, Integer idCategories) {
+        String sql = "select o from Product o where  o.categories.id = ?0";
+        return super.findManyPage(Product.class,sql,pageNumber,pageSize, idCategories );
+    }
+
+    @Override
+    public List<Product> findPageParam(Boolean existActive, int pageNumber, int pageSize, String title) {
+        String sql = "select o from Product o where o.title like ?0 or o.categories.title like ?1";
+        return super.findManyPage(Product.class,sql,pageNumber,pageSize,"%"+title+"%","%"+title+"%" );
+    }
+
+    @Override
+    public List<Product> findProductBetween(Double min, Double max) {
+        String sql = "select o from Product o where o.discount between ?0 and ?1";
+        List<Product> resultList = super.findMany(Product.class, sql, min, max);
+        if (resultList.size() <= 4) {
+            return resultList;
+        } else {
+            return resultList.subList(0, 4);
+        }
```

Comment:
@truongnguyenfpthcm2605 

Câu hỏi: Chỗ này sao lại có điều kiện lấy từ 0 đến 4 nhỉ?

Ngoài ra, đoạn code này phần else bị thừa, hãy sửa lại thành:

```
if (resultList.size() <= 4) {
  return resultList;
}
return resultList.subList(0, 4);
```

Reply:
Dạ phần này em sẽ lấy ra 4 sản phẩm có mức giá gần với sản phẩm đang xem

Reply:
@Override
    public List<Product> findProductBetweenDiscount(Double min, Double max) {
        String sql = "select o from Product o where o.discount between ?0 and ?1";
        List<Product> resultList = super.findMany(Product.class, sql, min, max);
        if (resultList.size() <= 4) {
            return resultList;
        }
        return resultList.subList(0, 4);
    }
    Em đã fix lại tên fuction và return